articles/tenntenn.re

```diff
@@ -0,0 +1,726 @@
+
+@<code>{pair.New(1, int64(2))}のように呼び出した場合、それぞれの実引数は@<code>{1}が型なしの定数、@<code>{int64(2)}が@<code>{int64}型の定数となります。
+型の単一化の最初のパスでは、@<code>{int64(2)}から@<code>{int64}型と型パラメタ@<code>{T}が対応することが分かります。
+最初のパスが終了した後、対応する型が見つからない型パラメタがないため、型の単一化は終了します。
+そのため、@<code>{pair.New(1, int64(2))}は@<code>{pair.New(int64)(1, int64(2))}と呼び出した場合と同様であると判断されます。
+
+@<code>{pair.New(1, 2.5)}のように呼び出した場合、それぞれの実引数は型なしの定数です。
+そのため、型の単一化の最初のパスでは処理されません。
+そこで、定数@<code>{1}と{2.5}のデフォルトの型である@<code>{int}型と@<code>{float64}型を用いて、2つめのパスが行われます。
+@<code>{int}型と@<code>{float64}型が型パラメタ@<code>{T}と対応しているため、矛盾が発生し型の単一化は失敗し、コンパイルエラーになります。
+
+型推論は型パラメタにコントラクトが指定しているかどうかにかかわらずに行われます。
+まず型推論が行われ型引数として暗黙的に指定される型が決定され、その後その型がコントラクトを満たしているか調べられます。
+
+型推論に成功したとしてもコンパイラは実引数が仮引数に割り当て可能かチェックします。
+これは型なしの定数を実引数として取った場合に限る話ではありません。
+
+型推論は非常に便利な機能ですが、型パラメタやコントラクトの設計に影響を与えるものではなく、単に利便性のために存在します。そのため、型パラメタやコントラクトに関する実装（ランタイムやコンパイラ）の初期段階では省略される可能性があります。
+
+関数呼び出し時の型推論のほかに@<list>{type_inference_composite_literal}のようなコンポジットリテラルにおける型推論を考えることもできます。しかし、このようなコードがどのくらいの頻度で必要になるのかは分からないため、実装されない可能性があります。
+
+//list[type_inference_composite_literal][コンポジットリテラルにおける型推論][go]{
+type Pair(type T) struct { f1, f2 T }
+var V = Pair{1, 2} // Pair(int){1, 2}のように推論される
+//}
+
+== 関数のインスタンス化
+
+Goでは関数を引数を渡さず参照することによって関数自体を値として扱うことができます（関数のインスタンス化）。
+型パラメタをもつ関数の場合、コンパイル時に型パラメタが解決されている必要があるため直接的にはインスタンス化できません。
+しかし、@<list>{PrintInts}のように型引数を指定することによって関数のインスタンス化を行うことが可能です。
+
+//list[PrintInts][型パラメタを持つ関数のインスタンス化][go]{
+// PrintIntsの型はfunc([]int)になる
+var PrintInts = Print(int)
+//}
+
+== 型アサーションと型スイッチ
+
+Goではインタフェースを実際の値にもとづき型変換を
+行うしくみとして型アサーションと型スイッチが用意されています。
+型パラメタで指定された型の値においても同様に型アサーションや型スイッチが行えます。
+
+たとえば、@<list>{typepram_typeassertion}は
+@<code>{ReadByte}関数の引数@<code>{r}に対して型アサーションを行っています。
+このとき、型パラメタ@<code>{T}に割り当てられる型はインタフェースである必要はありません。
+
+//list[typepram_typeassertion][型パラメタの型アサーション][go]{
+contract reader(T) {
+  T Read([]byte) (int, error)
+}
+
+func ReadByte(type T reader)(r T) (byte, error) {
+  if br, ok := r.(io.ByteReader); ok {
+    return br.ReadByte()
+  }
+  var b [1]byte
+  _, err := r.Read(b[:])
+  return b[0], err
+}
+//}
+
+== 型リテラルにおける型のインスタンス化
+
+型リテラルの末尾に型パラメタを用いた場合におけるインスタンス化を行う際、
+@<list>{typeliteral_typeinstantiating}のようにあいまいさ残る場合があります。
+
+//list[typeliteral_typeinstantiating][型リテラルの末尾に型パラメタを用いた場合におけるインスタンス化のあいまいさ][go]{
+x1 := []T(v1)
+x2 := []T(v2){}
+//}
+
+@<list>{typeliteral_typeinstantiating}の1行目は@<code>{v1}の@<code>{[]T}型へのキャストとなります。
+一方、2行目は@<code>{[]T(v2)}型のコンポジットリテラルとなります。
+このとき、@<code>{T}は型パラメタで型引数@<code>{v2}によってインスタンス化されています。
+
+このようなあいまいさを避けるため、型リテラルの末尾における型パラメタの
+インスタンス化は@<code>{[](T(v1))}のように@<code>{()}でくくる必要があります。
+
+@<code>{()}でくくらない場合、@<code>{[]T(x)}は@<code>{[](T(x))}ではなく、
+@<code>{([]T)(x)}と解釈されます。
+このルールは、型名で終わるスライス、配列、マップ、チャネル、関数の型リテラルの場合のみ適用されます。
+
+== 無名関数における型パラメタの利用
+== リフレクション
+== コントラクトの詳細
+=== メソッド
+=== 演算子
+=== コントラクト内の型
+=== 制約の連言と選言
+=== 型の集約
+=== 型パラメタの集約
+=== コントラクトにおける型についての観察（？）
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "？". (ja-technical-writing/no-exclamation-question-mark)